volatility3/framework/symbols/linux/utilities/modules.py

```diff
@@ -534,3 +539,98 @@ def validate_alignment_patterns(
             True if all the addresses meet the alignment
         """
         return all(addr % address_alignment == 0 for addr in addresses)
+
+
+class ModuleGathererLsmod(ModuleGathererInterface):
+    """
+    Gathers modules from the main kernel list
+    """
+
+    @classmethod
+    def gather_modules(
+        cls, context: interfaces.context.ContextInterface, kernel_module_name: str
+    ) -> ModuleGathererInterface.gatherer_return_type:
+        yield from Modules.list_modules(context, kernel_module_name)
+
+
+class ModuleGathererSysFs(ModuleGathererInterface):
+    """
+    Gathers modules from the sysfs /sys/modules objects
+    """
+
+    @classmethod
+    def gather_modules(
+        cls, context: interfaces.context.ContextInterface, kernel_module_name: str
+    ) -> ModuleGathererInterface.gatherer_return_type:
+        kernel = context.modules[kernel_module_name]
+
+        sysfs_modules: dict = Modules.get_kset_modules(context, kernel_module_name)
+
+        for m_offset in sysfs_modules.values():
+            yield kernel.object(object_type="module", offset=m_offset, absolute=True)
+
+
+class ModuleGathererScanner(ModuleGathererInterface):
+    """
+    Gathers modules by scanning memory
+    """
+
+    @classmethod
+    def gather_modules(
+        cls, context: interfaces.context.ContextInterface, kernel_module_name: str
+    ) -> ModuleGathererInterface.gatherer_return_type:
+        modules_memory_boundaries = Modules.get_modules_memory_boundaries(
+            context, kernel_module_name
+        )
+
+        # Send in an empty list to not filter on any modules
+        yield from Modules.get_hidden_modules(
+            context=context,
+            vmlinux_module_name=kernel_module_name,
+            known_module_addresses=[],
+            modules_memory_boundaries=modules_memory_boundaries,
+        )
+
+
+class ModuleGathererKernel(ModuleGathererInterface):
+    """
+    Creates a ModuleInfo instance for the kernel so that plugins
+    can determine when function pointers reference the kernel
+    """
+
+    @classmethod
+    def gather_modules(
+        cls, context: interfaces.context.ContextInterface, kernel_module_name: str
+    ) -> ModuleGathererInterface.gatherer_return_type:
+        """
+        Returns a ModuleInfo instance that encodes the kernel
+        This is required to map function pointers to the kerenl executable
+        """
+        kernel = context.modules[kernel_module_name]
+
+        address_mask = context.layers[kernel.layer_name].address_mask
+
+        start_addr = kernel.object_from_symbol("_text")
+        start_addr = start_addr.vol.offset & address_mask
+
+        end_addr = kernel.object_from_symbol("_etext")
+        end_addr = end_addr.vol.offset & address_mask
+
+        yield ModuleInfo(start_addr, constants.linux.KERNEL_NAME, start_addr, end_addr)
+
+
+class ModuleGatherers(interfaces.configuration.VersionableInterface):
```

Comment:
I'm... not sure this is useful?  I'd just keep the list versioned inside the `Modules` component?  I might also make each of the scanners live inside it, so that they inherit the versioning too?  Otherwise we're just versioning the list of all of them, but not each of the individual gatherers...

Reply:
But then we would need to version bump Modules on changes, which only plugins that look for rootkit signs care about that, but we would have to version bump ones that don't care about rootkits also. This breaks the versioning out so only plugins wanting the functionality have to care.

Reply:
Yeah, but just like very thing, we'd work hard to only make additive bumps (which then won't require the plugins to change).  This way, a change here isn't tracked, only a change in the default framework collection of trackers is versioned, and people will need to check the version of both the `ModuleGatherers` list *and* of the `Modules` class, when they're kinda bound together?

Reply:
Also, I really don't like everything shoved in one class. Having versioned classes as small as makes sense (grouped functionality) means plugins can make specific requirements on just what they need and we avoid the endless version bumps for plugins of unrelated work.

Reply:
True, there comes a point where you're needing everyone to version everything.  At the very least the gatherers themselves need to be under *something* versioned, otherwise technically they're linked to the framework version...  5:S

Reply:
Do they not inherit the version from the interface? I was pretty certain it worked that way, so I designed it where, if the interface ever version bumps, then the gatherers have to adjust.

Reply:
yes, that makes sense, that will now be a part of the update where each gatherer gets versioned.